addons/payment/models/payment_transaction.py

```diff
@@ -631,81 +631,87 @@ def _process_notification_data(self, notification_data):
+        allowed_states = ['draft', 'pending', 'authorized', 'error']
         target_state = 'done'
-        txs_to_process = self._update_state(allowed_states, target_state, state_message)
+        txs_to_process = self._update_state(allowed_states, extra_allowed_states, target_state, state_message)
         txs_to_process._log_received_message()
         return txs_to_process
 
-    def _set_canceled(self, state_message=None):
+    def _set_canceled(self, state_message=None, extra_allowed_states=None):
         """ Update the transactions' state to `cancel`.
 
         :param str state_message: The reason for setting the transactions in the state `cancel`.
+        :param list extra_allowed_states: Additional allowed states specific to a provider on some flows
         :return: The updated transactions.
         :rtype: recordset of `payment.transaction`
         """
-        allowed_states = ('draft', 'pending', 'authorized', 'done')  # 'done' for Authorize refunds.
+        allowed_states = ['draft', 'pending', 'authorized']
         target_state = 'cancel'
-        txs_to_process = self._update_state(allowed_states, target_state, state_message)
+        txs_to_process = self._update_state(allowed_states, extra_allowed_states, target_state, state_message)
         # Cancel the existing payments.
         txs_to_process._log_received_message()
         return txs_to_process
 
-    def _set_error(self, state_message):
+    def _set_error(self, state_message, extra_allowed_states=None):
         """ Update the transactions' state to `error`.
 
         :param str state_message: The reason for setting the transactions in the state `error`.
+        :param list extra_allowed_states: Additional allowed states specific to a provider on some flows
         :return: The updated transactions.
         :rtype: recordset of `payment.transaction`
         """
-        allowed_states = ('draft', 'pending', 'authorized', 'done')  # 'done' for Stripe refunds.
+        allowed_states = ['draft', 'pending', 'authorized']
         target_state = 'error'
-        txs_to_process = self._update_state(allowed_states, target_state, state_message)
+        txs_to_process = self._update_state(allowed_states, extra_allowed_states, target_state, state_message)
         txs_to_process._log_received_message()
         return txs_to_process
 
-    def _update_state(self, allowed_states, target_state, state_message):
+    def _update_state(self, allowed_states, extra_allowed_states, target_state, state_message):
         """ Update the transactions' state to the target state if the current state allows it.
 
         If the current state is the same as the target state, the transaction is skipped and a log
         with level INFO is created.
 
-        :param tuple[str] allowed_states: The allowed source states for the target state.
+        :param list allowed_states: The allowed source states for the target state.
```

Comment:
```suggestion
        :param list[str] allowed_states: The allowed source states for the target state.
```
You can keep the type of the inside elements. Could also be interesting to add this for all the `extra_allowed_states` now that I though of it :thinking: